Biomes Structural Conversion

ZoltanTheDM opened the pull request:
I know this is a large commit, and I am sorry about that. But I did my best to not change the behavior of anything and just convert the series of arrays that defined biomes and turn them into a class.

I hope this helps make the project a little more organized.

Azgaar commented:
I'm not against classes, but in this case it provides nothing and not save in terms of possible bugs. The current code looks like:
`const icon = getBiomeIcon(i, biomesData.icons[b]);`

A proposed changed:
`const icon = getBiomeIcon(i, biomesData.biomeList[b].icons.probability);`

So it make the code even more complex.

ZoltanTheDM commented:
Good point there were some places I could make it more readable. There are some places that are still somewhat complicated but those should be improved with a future PR

Azgaar commented:
I believe we can exclude biomesList and add data as pack.biomes to inline it with rivers and other data.

Azgaar commented:
But in any case it will not make the code significantly better and can cause a lot of issues

ZoltanTheDM commented:
> 
> 
> I believe we can exclude biomesList and add data as pack.biomes to inline it with rivers and other data.

Having the biomes in the pack is a good idea, I noticed it was rather inconsistent not being there.


> 
> 
> But in any case it will not make the code significantly better and can cause a lot of issues

I agree that this is a major change that requires testing, however it also make the code more readable and usable which helps out a new coder to the project. This is your project so I won't press the issue. But I am working on an improvement that will also turn Cell data into a class structure. These 2 combined will make a huge improvement in usability. Which is great for my future ideas for this project.

Azgaar commented:
Biomes data is kinda separate and I'm ok to handle it in OOP approach. But cells data is much more complex and it's stored in typed arrays, which the fastest way to keep the data in js. I'm not sure that  development paradigm change is a good idea at this point of time.

ZoltanTheDM commented:
I’ll do OOP as an experiment.

However I would like to walk you through the following thought experiment for why I don’t think typed arrays are the fastest in this instance.

First some nitty gritty. The (simplified) price of an operation: (disclaimer I don’t know exactly how JavaScript is actually parsed and executed but I can make educated guesses)

*References* should be just an address lookup and should be completed with just a few cycles as they are just pointers. I’ll refer to this price as **R**

*Symbol lookup*, Since JavaScript is an interpreted language it must lookup what a name means. I believe it uses hash tables which should be relatively fast but not free. This includes member variables. ie `a.b.c` is 3 lookups, or **L** * 3. Notably each one must include a reference.

*Numerically Index Arrays* should just be a `pointer + index * sizeof object` then a reference a very fast operation. (price to be known as **A**). TypedArray are fast because that is the end of it, you have your resulting number. Non-typed arrays will have to go through an Extra reference to reach the final object (another R)

Armed with this information I would like to go through 2 thought experiments

## Find the biome habitability of a burg given its ID

#### Current

`biomesData[cells.biome[burgs[ID].cell]].habitability`
Operational Price: 7L + 3A + 1R

#### By reference (OOP)

`burgs[ID].cell.biome.habitability`
Operational Price: 5L + 1A + 4R

That examples is somewhat contrived operation but I would point out that OOP is also easier to follow

## forEach Example

A more common operation is for every cell in a list perform a change on an element

#### Current

`cellIDList.forEach(i => cells.element[i] = change())`
Price of single cycle of operation: 2A + 3L

#### OOP

`cellList.forEach(cell => cell.element = change())`
Price of single cycle of operation: 1A + 2L + 2R

## My Conclusion

Passing by reference is faster and easier to follow. If I am wrong in my conclusion please tell me as I am interested in learning.

Azgaar commented:
JavaScript can be pretty comlex and unobvious and you can prove the performance only using real tests, not theoretical examples. But even if objects are not slow, they need much more memory to be stored. Typed arrays is a reccomentded way to store big amout (MBs) of data. And in any case we use master objects like pack and grid, but actuall data is in typed arrays. 

It's easy to work and store and I don't see an issue with it.

ZoltanTheDM commented:
I would like to shelve the OOP conversation of cells for now as it has gotten far off topic from this PR. There is little more to be said without concrete examples. Thanks for thinking about it however.

Azgaar commented:
Thank you for the PR. Raised question are big too generic, I'm not ready to change the data structure when we have so many bugs to fix and features to implement :)

ZoltanTheDM commented:
I have fixed up this PR a little bit. I noticed that saving and loading was broken so I fixed that. I also moved biomes to be in the pack to be consistent with other data structures.

So far I have tested this branch by adding and removing custom biomes, painting biomes, population and habitability changes seem to work. Saving and loading seems to be unaffected. I would encourage you to try and break it but it seems ready if you are willing to merge.

Performance doesn't seem to be meaningfully impacted.

Azgaar commented:
That's a significant change that can easily brake the system. There are thousands of use cases in the tool and no automated unit and integration tests created. I'm not ready to merge it without going through alpha. 
And I still think that even the change is cool and I like classes, the practical gain is minimal.

Azgaar commented:
Closing for now as the change to data structure is pretty complex and would require extensive testing. Generally it would be better to start with minor functional enhancements / bug fixes / separate functions refactoring, not from so significant changes. We can get back to the current change later when major refactoring is in planned.
